crates/lang/codegen/src/generator/storage_item.rs

```diff
@@ -0,0 +1,249 @@
+                quote! {
+                    #vis struct #struct_ident #generics {
+                        #(#fields),*
+                    }
+                }
+            }
+        }
+    }
+
+    fn generate_enum(&self, enum_item: DataEnum) -> TokenStream2 {
+        let item = self.item;
+        let enum_ident = item.ident();
+        let vis = item.vis();
+        let generics = item.generics();
+        let salt = item.salt();
+
+        let variants = enum_item.variants.into_iter().map(|variant| {
+            let attrs = variant.attrs;
+            let variant_ident = &variant.ident;
+            let discriminant = if let Some((eq, expr)) = variant.discriminant {
+                quote! { #eq #expr}
+            } else {
+                quote! {}
+            };
+
+            let fields: Vec<_> = variant
+                .fields
+                .iter()
+                .enumerate()
+                .map(|(i, field)| {
+                    convert_into_storage_field(
+                        enum_ident,
+                        Some(variant_ident),
+                        &salt,
+                        i,
+                        field,
+                    )
+                })
+                .collect();
+
+            let fields = match variant.fields {
+                Fields::Named(_) => quote! { { #(#fields),* } },
+                Fields::Unnamed(_) => quote! { ( #(#fields),* ) },
+                Fields::Unit => quote! {},
+            };
+
+            quote! {
+                #(#attrs)*
+                #variant_ident #fields #discriminant
+            }
+        });
+
+        quote! {
+            #vis enum #enum_ident #generics {
+                #(#variants),*
+            }
+        }
+    }
+
+    fn generate_union(&self, union_item: DataUnion) -> TokenStream2 {
+        let item = self.item;
+        let union_ident = item.ident();
+        let vis = item.vis();
+        let generics = item.generics();
+        let salt = item.salt();
+
+        let fields = union_item
+            .fields
+            .named
+            .iter()
+            .enumerate()
+            .map(|(i, field)| {
+                convert_into_storage_field(union_ident, None, &salt, i, field)
+            });
+
+        quote! {
+            #vis union #union_ident #generics {
+                #(#fields),*
+            }
+        }
+    }
+
+    fn generate_type_check(&self) -> TokenStream2 {
+        let fields = self
+            .item
+            .all_used_types()
+            .into_iter()
+            .enumerate()
+            .map(|(i, ty)| {
+                let field_name = format_ident!("field_{}", i);
+                let span = ty.span();
+                quote_spanned!(span =>
+                    #field_name: #ty
+                )
+            });
+        let generics = self.item.generics();
+        let salt = self.item.salt();
+
+        quote! {
+            const _: () = {
+                struct Check #generics {
```

Comment:
Why's this `Check` struct required?

Reply:
To return a user-friendly error if the user uses the wrong types. Without it, we will return that some invalid types are not implemented `StorageHint`.